File: android/app/src/main/java/com/descope/deviceauthdemo/network/DeviceAuthService.kt
package com.descope.deviceauthdemo.network

import com.descope.deviceauthdemo.Config
import com.descope.deviceauthdemo.model.DeviceAuthorizationResponse
import com.descope.deviceauthdemo.model.DeviceFlowPollError
import com.descope.deviceauthdemo.model.DeviceTokenResponse
import com.descope.deviceauthdemo.model.optStringOrNull
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import okhttp3.FormBody
import okhttp3.OkHttpClient
import okhttp3.Request
import org.json.JSONObject
import java.io.IOException

/**
 * Talks to Descope's OAuth 2.0 Device Authorization Grant endpoints.
 * Reference: https://docs.descope.com/auth-methods/device-auth
 */
class DeviceAuthService(
    private val client: OkHttpClient = OkHttpClient()
) {

    /** The outcome of a single poll against the token endpoint. */
    sealed class PollResult {
        data class Success(val token: DeviceTokenResponse) : PollResult()
        data class Pending(val reason: DeviceFlowPollError) : PollResult() // authorization_pending or slow_down
        data class TerminalFailure(val reason: DeviceFlowPollError) : PollResult() // access_denied or expired_token
    }

    class DeviceAuthException(message: String) : IOException(message)

    /** Step 1: request a `device_code` / `user_code` pair. */
    suspend fun startDeviceAuthorization(): DeviceAuthorizationResponse = withContext(Dispatchers.IO) {
        if (!Config.isConfigured) {
            throw DeviceAuthException("Set your Descope Project ID in Config.kt before running this demo.")
        }

        val body = FormBody.Builder()
            .add("client_id", Config.clientId)
            .add("scope", Config.scope)
            .build()

        val request = Request.Builder()
            .url(Config.deviceAuthorizationUrl)
            .post(body)
            .build()

        client.newCall(request).execute().use { response ->
            val responseBody = response.body?.string().orEmpty()
            if (!response.isSuccessful) {
                throw DeviceAuthException(describeError(response.code, responseBody))
            }
            DeviceAuthorizationResponse.fromJson(JSONObject(responseBody))
        }
    }

    /** Descope's general error response shape, e.g. `{"errorCode", "errorDescription", "errorMessage", "message"}`. */
    private fun describeError(status: Int, responseBody: String): String {
        val message = try {
            val json = JSONObject(responseBody)
            json.optStringOrNull("message")
                ?: json.optStringOrNull("errorMessage")
                ?: json.optStringOrNull("errorDescription")
        } catch (e: Exception) {
            null
        }
        return if (message != null) "$message (status $status)" else "Unexpected server response (status $status)"
    }

    /**
     * Step 2: poll the token endpoint with the `device_code` grant.
     * Call this on an interval (see [DeviceAuthorizationResponse.interval])
     * until it returns [PollResult.Success] or a [PollResult.TerminalFailure].
     */
    suspend fun pollForToken(deviceCode: String): PollResult = withContext(Dispatchers.IO) {
        val body = FormBody.Builder()
            .add("grant_type", "urn:ietf:params:oauth:grant-type:device_code")
            .add("device_code", deviceCode)
            .add("client_id", Config.clientId)
            .build()

        val request = Request.Builder()
            .url(Config.tokenUrl)
            .post(body)
            .build()

        client.newCall(request).execute().use { response ->
            val responseBody = response.body?.string().orEmpty()
            val json = try {
                JSONObject(responseBody)
            } catch (e: Exception) {
                null
            }

            if (response.isSuccessful && json != null) {
                return@use PollResult.Success(DeviceTokenResponse.fromJson(json))
            }

            val errorCode = json?.optString("error")
            val pollError = DeviceFlowPollError.from(errorCode)
            return@use when (pollError) {
                DeviceFlowPollError.AUTHORIZATION_PENDING,
                DeviceFlowPollError.SLOW_DOWN -> PollResult.Pending(pollError)

                DeviceFlowPollError.ACCESS_DENIED,
                DeviceFlowPollError.EXPIRED_TOKEN,
                DeviceFlowPollError.UNKNOWN -> {
                    if (pollError == DeviceFlowPollError.UNKNOWN && json == null) {
                        throw DeviceAuthException(describeError(response.code, responseBody))
                    }
                    PollResult.TerminalFailure(pollError)
                }
            }
        }
    }
}

File: ios/DeviceAuthDemo/Services/DeviceAuthService.swift
import Foundation

/// Talks to Descope's OAuth 2.0 Device Authorization Grant endpoints.
/// Reference: https://docs.descope.com/auth-methods/device-auth
struct DeviceAuthService {

    enum ServiceError: Error, LocalizedError {
        case notConfigured
        case transport(Error)
        case unexpectedStatus(Int, Data)
        case decoding(Error)

        var errorDescription: String? {
            switch self {
            case .notConfigured:
                return "Set your Descope Project ID in DescopeConfig.swift before running this demo."
            case .transport(let error):
                return "Network error: \(error.localizedDescription)"
            case .unexpectedStatus(let status, let data):
                if let descopeError = try? JSONDecoder().decode(DescopeErrorBody.self, from: data),
                   let message = descopeError.message ?? descopeError.errorMessage ?? descopeError.errorDescription {
                    return "\(message) (status \(status))"
                }
                return "Unexpected server response (status \(status))"
            case .decoding:
                return "Couldn't parse the server's response"
            }
        }
    }

    /// Descope's general error response shape (distinct from the RFC 8628
    /// `{error, error_description}` shape used specifically by the token
    /// endpoint's pending/terminal states).
    private struct DescopeErrorBody: Decodable {
        let errorCode: String?
        let errorDescription: String?
        let errorMessage: String?
        let message: String?
    }

    /// The outcome of a single poll against the token endpoint.
    enum PollResult {
        case success(DeviceTokenResponse)
        case pending(DeviceFlowPollError) // authorization_pending or slow_down
        case terminalFailure(DeviceFlowPollError) // access_denied or expired_token
    }

    private let session: URLSession

    init(session: URLSession = .shared) {
        self.session = session
    }

    /// Step 1: request a `device_code` / `user_code` pair.
    func startDeviceAuthorization() async throws -> DeviceAuthorizationResponse {
        guard DescopeConfig.isConfigured else {
            throw ServiceError.notConfigured
        }

        var request = URLRequest(url: DescopeConfig.deviceAuthorizationURL)
        request.httpMethod = "POST"
        request.setValue("application/x-www-form-urlencoded", forHTTPHeaderField: "Content-Type")

        let bodyParams = [
            "client_id": DescopeConfig.clientId,
            "scope": DescopeConfig.scope
        ]
        request.httpBody = formEncode(bodyParams)

        let (data, response) = try await perform(request)
        guard let http = response as? HTTPURLResponse, (200...299).contains(http.statusCode) else {
            let status = (response as? HTTPURLResponse)?.statusCode ?? -1
            throw ServiceError.unexpectedStatus(status, data)
        }

        do {
            return try JSONDecoder().decode(DeviceAuthorizationResponse.self, from: data)
        } catch {
            throw ServiceError.decoding(error)
        }
    }

    /// Step 2: poll the token endpoint with the `device_code` grant.
    /// Call this on an interval (see `DeviceAuthorizationResponse.interval`)
    /// until it returns `.success` or a `.terminalFailure`.
    func pollForToken(deviceCode: String) async throws -> PollResult {
        var request = URLRequest(url: DescopeConfig.tokenURL)
        request.httpMethod = "POST"
        request.setValue("application/x-www-form-urlencoded", forHTTPHeaderField: "Content-Type")

        let bodyParams = [
            "grant_type": "urn:ietf:params:oauth:grant-type:device_code",
            "device_code": deviceCode,
            "client_id": DescopeConfig.clientId
        ]
        request.httpBody = formEncode(bodyParams)

        let (data, response) = try await perform(request)
        let status = (response as? HTTPURLResponse)?.statusCode ?? -1

        if (200...299).contains(status) {
            do {
                let token = try JSONDecoder().decode(DeviceTokenResponse.self, from: data)
                return .success(token)
            } catch {
                throw ServiceError.decoding(error)
            }
        }

        // Non-2xx: expect an RFC 8628 error body naming what to do next.
        guard let errorBody = try? JSONDecoder().decode(DeviceTokenErrorResponse.self, from: data) else {
            throw ServiceError.unexpectedStatus(status, data)
        }

        let pollError = DeviceFlowPollError(rawValue: errorBody.error)
        switch pollError {
        case .authorizationPending, .slowDown:
            return .pending(pollError)
        case .accessDenied, .expiredToken, .unknown:
            return .terminalFailure(pollError)
        }
    }

    private func perform(_ request: URLRequest) async throws -> (Data, URLResponse) {
        do {
            return try await session.data(for: request)
        } catch {
            throw ServiceError.transport(error)
        }
    }

    private func formEncode(_ params: [String: String]) -> Data {
        let pairs = params.map { key, value -> String in
            let allowed = CharacterSet.urlQueryAllowed.subtracting(CharacterSet(charactersIn: "+&="))
            let encodedKey = key.addingPercentEncoding(withAllowedCharacters: allowed) ?? key
            let encodedValue = value.addingPercentEncoding(withAllowedCharacters: allowed) ?? value
            return "\(encodedKey)=\(encodedValue)"
        }
        return pairs.joined(separator: "&").data(using: .utf8) ?? Data()
    }
}
